File: python/walking.py
#==============================================================================#
#
# Module for getting Google Maps walking directions to and from stops.
#
#==============================================================================#

from urllib2      import urlopen
from json         import load

"""
walk_to(start, stop)

Parameters:
    - start: a (lat,lng) tuple
    - stop:  a TransLoc shuttle stop JSON data item

Returns:
    - directions: a list of HTML directions
"""
def walk_to(start, stop):
    url = 'http://maps.googleapis.com/maps/api/directions/json?'
    origin = 'origin=' + str(start[0]) + ',' + str(start[1])
    destination = '&destination=' + \
        str(stop['location']['lat']) + ',' + str(stop['location']['lng'])
    url += origin
    url += destination
    url += '&sensor=true'
    url += '&mode=walking'

    response = urlopen(url)
    json = load(response)

    directions = []
    for step in json['routes'][0]['legs'][0]['steps']:
        directions.append(step['html_instructions'])
    return directions

"""
walk_from(stop, dest):

Parameters:
    - stop: a TransLoc shuttle stop JSON data item
    - dest: a (lat,lng) tuple

Returns:
    - directions: a list of HTML directions
"""
def walk_from(stop, dest):

    url = 'http://maps.googleapis.com/maps/api/directions/json?'
    origin = 'origin=' + \
        str(stop['location']['lat']) + ',' + str(stop['location']['lng'])
    destination = '&destination=' + str(dest[0]) + ',' + str(dest[1])
    url += origin
    url += destination
    url += '&sensor=true'
    url += '&mode=walking'

    response = urlopen(url)
    json = load(response)

    directions = []
    for step in json['routes'][0]['legs'][0]['steps']:
        directions.append(step['html_instructions'])
    return directions
    

File: java/Walking.java
// Class for getting Google Maps walking directions

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.UnsupportedEncodingException;
import java.net.URL;
import java.net.URLConnection;
import java.security.MessageDigest;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;
import java.util.Iterator;
import java.lang.Math;

import org.json.*;


public class Walking {

    private static String GOOGLEMAPS = "http://maps.googleapis.com/maps/api/directions/json?";

	public static ArrayList<String> walkTo(double lat, double lng, HashMap stop) throws IOException, JSONException {
        String ORIGIN = "origin=" + lat + "," + lng;
        String DEST   = "&destination=" + ((JSONObject)stop.get("location")).getString("lat")
                                  + "," + ((JSONObject)stop.get("location")).getString("lng");
        String INFO   = "&sensor=true&mode=walking";

        URL url = new URL(GOOGLEMAPS + ORIGIN + DEST + INFO);
        URLConnection connection = url.openConnection();

        String line;
        StringBuilder builder = new StringBuilder();
        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
        while((line = reader.readLine()) != null) {
            builder.append(line);
        }

        ArrayList<String> directions = new ArrayList<String>();
        JSONObject jObj = new JSONObject(builder.toString());
        JSONArray steps = jObj.getJSONArray("routes").getJSONObject(0).getJSONArray("legs").getJSONObject(0).getJSONArray("steps");

        for (int i=0; i < steps.length(); i++)
            directions.add(steps.getJSONObject(i).getString("html_instructions"));

        return directions;
    }

    public static ArrayList<String> walkFrom(HashMap stop, double lat, double lng) throws IOException, JSONException {
        String DEST   = "&destination=" + ((JSONObject)stop.get("location")).getString("lat")
                                  + "," + ((JSONObject)stop.get("location")).getString("lng");
        String ORIGIN = "origin=" + lat + "," + lng;
        String INFO   = "&sensor=true&mode=walking";

        URL url = new URL(GOOGLEMAPS + ORIGIN + DEST + INFO);
        URLConnection connection = url.openConnection();

        String line;
        StringBuilder builder = new StringBuilder();
        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
        while((line = reader.readLine()) != null) {
            builder.append(line);
        }

        ArrayList<String> directions = new ArrayList<String>();
        JSONObject jObj = new JSONObject(builder.toString());
        JSONArray steps = jObj.getJSONArray("routes").getJSONObject(0).getJSONArray("legs").getJSONObject(0).getJSONArray("steps");

        for (int i=0; i < steps.length(); i++)
            directions.add(steps.getJSONObject(i).getString("html_instructions"));
        
        return directions;
    }

	public static void main(String[] args) throws IOException, JSONException {
        ArrayList<String> to = walkTo(41.791393,-87.599776,TransLocAPI.getStops("100").get(0));
        System.out.println(to);
        ArrayList<String> from = walkFrom(TransLocAPI.getStops("100").get(0),41.791393,-87.599776);
        System.out.println(from);
		return;		
	}
}






